libmspack/mspack/mspack-version.h.in

```diff
@@ -0,0 +1,32 @@
+#ifndef MSPACK_VER_H
+#define MSPACK_VER_H
+
+/**
+ * @macro
+ * Version string of the mspack project release.
+ */
+#define MSPACK_VERSION "@PROJECT_VERSION@"
```

Comment:
Disclaimer: I don't pretend to be an expert at library package management stuff. I mostly just grumble about it (especially when it has to do with freebsd 😀 ).

As far as I know there is no pkg-config on Windows (native - not counting MingW/Cygwin), on macOS, or on a few other operating systems. Granted on Windows applications are generally deployed with their dependencies. Afaik, no-one installs DLLs to a common location to be shared between different applications. I have to wonder if Microsoft is planning to change that with the new "winget" package manager 😕. 

In any case, I don't think pkg-config can be relied on to exist everywhere.  I think version macros are most reliable way for an app developer to support multiple versions of a library.

Reply:
I'm not a packaging expert either, but I see a lot of value in it, which is why I use Cygwin, macports and so on. Hopefully Microsoft will too with ~AppGet~WinGet.

To me, it looks like the CMake way of doing things should involve find_package. The value in find_package is:

* it will import the config from the package you want to use, so both library and caller have the same build settings 
* if you don't have the required version, cmake will say this unambigiously

I feel that trying to do package management in header files is too late (header files can't make the build system import necessary config/defines), and using headers to fail a build by checking version numbers is doing the packaging/build system's job for it.

From the other direction, I also don't see the value of version numbers in headers to control conditional code execution, because libmspack is always backwards compatible, and provides runtime feature checking.

Whatever version of <mspack.h> you have, you can compile code that uses all fields of all structures it defines. To be sure your code doesn't crash or access OOB memory, you should check the appropriate `mspack_version()` before relying on using something that only exists in V2, V3 of a specific API. A version number in a header can't give you this assurance, because header files can't control what version of library is linked or what .DLL is put in the same directory as the executable.

The only reason I can think of for including the *packaging* version number in libmspack is to provide it to end users as assurance, e.g. sysadmins wanting to know their libmspack-consuming-program isn't using a vulnerable version of libmspack. Even then, I don't think it should be in a header, because it could bear false witness. It would have to be a function in the public API, to be assured it came from the library/DLL whose code will really run.

Are there other uses of version numbers in library headers that I'm not considering?